pkg/cmd/cli/cmd/new/newapp.go

```diff
@@ -0,0 +1,383 @@
+			}
+			unknown := config.addArguments(args)
+			if len(unknown) != 0 {
+				glog.Fatalf("Did not recognize the following arguments: %v", unknown)
+			}
+			if err := config.Run(f, out, c.Help); err != nil {
+				if errs, ok := err.(errlist); ok {
+					if len(errs.Errors()) == 1 {
+						err = errs.Errors()[0]
+					}
+				}
+				if usage, ok := err.(UsageError); ok {
+					glog.Fatal(usage.UsageError(c.CommandPath()))
+				}
+				glog.Fatalf("Error: %v", err)
+			}
+		},
+	}
+
+	cmd.Flags().Var(&config.SourceRepositories, "code", "Source code to use to build this application.")
+	cmd.Flags().VarP(&config.ImageStreams, "image", "i", "Name of an OpenShift image repository to use in the app.")
+	cmd.Flags().Var(&config.DockerImages, "docker-image", "Name of a Docker image to include in the app.")
+	cmd.Flags().Var(&config.Groups, "group", "Indicate components that should be grouped together as <comp1>+<comp2>.")
+	cmd.Flags().VarP(&config.Environment, "env", "e", "Specify key value pairs of environment variables to set into each container.")
+	cmd.Flags().StringVar(&config.TypeOfBuild, "build", "", "Specify the type of build to use if you don't want to detect (docker|source)")
+	return cmd
+}
+
+type UsageError interface {
+	UsageError(commandName string) string
+}
+
+// TODO: replace with upstream converting [1]error to error
+type errlist interface {
+	Errors() []error
+}
+
+// addArguments converts command line arguments into the appropriate bucket based on what they look like
+func (c *newAppConfig) addArguments(args []string) []string {
+	unknown := []string{}
+	for _, s := range args {
+		switch {
+		case cmdutil.IsEnvironmentArgument(s):
+			c.Environment = append(c.Environment, s)
+		case isPossibleSourceRepository(s):
+			c.SourceRepositories = append(c.SourceRepositories, s)
+		case isComponentReference(s):
+			c.Components = append(c.Components, s)
+		default:
+			if len(s) == 0 {
+				break
+			}
+			unknown = append(unknown, s)
+		}
+	}
+	return unknown
+}
+
+// validate converts all of the arguments on the config into references to objects, or returns an error
+func (c *newAppConfig) validate() (ComponentReferences, []*SourceRepository, cmdutil.Environment, error) {
+	b := &ReferenceBuilder{}
+	for _, s := range c.SourceRepositories {
+		b.AddSourceRepository(s)
+	}
+	b.AddImages(c.DockerImages, func(input *ComponentInput) ComponentReference {
+		input.Argument = fmt.Sprintf("--docker-image=%q", input.From)
+		input.Resolver = c.localDockerResolver
+		return input
+	})
+	b.AddImages(c.ImageStreams, func(input *ComponentInput) ComponentReference {
+		input.Argument = fmt.Sprintf("--image=%q", input.From)
+		input.Resolver = c.imageStreamResolver
+		return input
+	})
+	b.AddImages(c.Components, func(input *ComponentInput) ComponentReference {
+		input.Resolver = PerfectMatchWeightedResolver{
+			WeightedResolver{Resolver: c.imageStreamResolver, Weight: 0.0},
+			WeightedResolver{Resolver: c.localDockerResolver, Weight: 0.0},
+		}
```

Comment:
Yes - when we do search, I want the searchers to correctly order things based on this order:

* image streams in your namespace, templates in your namespace
* image streams in your search order, templates in your search order (and search order should weight them)
* docker client locally
* docker registry
* other things in the future like an openshift searcher

----- Original Message -----
> > +	}
> > +	b.AddImages(c.DockerImages, func(input *ComponentInput)
> > ComponentReference {
> > +		input.Argument = fmt.Sprintf("--docker-image=%q", input.From)
> > +		input.Resolver = c.localDockerResolver
> > +		return input
> > +	})
> > +	b.AddImages(c.ImageStreams, func(input *ComponentInput)
> > ComponentReference {
> > +		input.Argument = fmt.Sprintf("--image=%q", input.From)
> > +		input.Resolver = c.imageStreamResolver
> > +		return input
> > +	})
> > +	b.AddImages(c.Components, func(input *ComponentInput) ComponentReference
> > {
> > +		input.Resolver = PerfectMatchWeightedResolver{
> > +			WeightedResolver{Resolver: c.imageStreamResolver, Weight: 0.0},
> > +			WeightedResolver{Resolver: c.localDockerResolver, Weight: 0.0},
> > +		}
> 
> Do we have a use case in mind for resolvers of different weight? Seems
> unnecessary complexity if we don't
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/668/files#r23224251